scripts/leapfrog/generate_resources.py

```diff
@@ -0,0 +1,322 @@
+#! /usr/bin/env python
+from argparse import ArgumentParser
+from ipaddress import IPv4Network
+from itertools import chain, izip, repeat, islice
+from json import dumps
+import os
+from time import sleep
+
+from keystoneauth1.identity import v3
+from keystoneauth1 import session as session_
+from cinderclient import client as cinder_client
+from glanceclient import client as glance_client
+from neutronclient.v2_0 import client as neutron_client
+from novaclient import client as nova_client
+
+CLIENT_VERSION = "2"
+
+
+def get_conf():
+    parser = ArgumentParser()
+    parser.add_argument("--servers", type=int, default=1)
+    parser.add_argument("--volumes", type=int, default=0)
+    parser.add_argument("--networks", type=int, default=1)
+    parser.add_argument("--infinite-quotas", action="store_true")
+    args = parser.parse_args()
+    if args.servers > args.networks:
+        raise ValueError(
+            "Each server must have at least one network and so the number of "
+            "networks must be greater than or equal to the number of servers."
+        )
+    if args.servers < 1:
+        raise ValueError(
+            "The minimum number of servers that can be created is 1."
+        )
+
+    conf = {
+        "auth": {},
+        "infinite_quotas": args.infinite_quotas,
+        "resources": {},
+    }
+    conf["auth"]["endpoint_type"] = os.environ["OS_ENDPOINT_TYPE"]
+    conf["auth"]["username"] = os.environ["OS_USERNAME"]
+    conf["auth"]["password"] = os.environ["OS_PASSWORD"]
+    conf["auth"]["project_name"] = os.environ["OS_PROJECT_NAME"]
+    conf["auth"]["auth_url"] = os.environ["OS_AUTH_URL"]
+    conf["auth"]["user_domain_id"] = os.environ["OS_USER_DOMAIN_NAME"]
+    conf["auth"]["project_domain_id"] = os.environ["OS_PROJECT_DOMAIN_NAME"]
+
+    conf["resources"]["servers"] = {
+        "count": args.servers,
+        "flavor": "m1.tiny",
+        "image": "cirros",
+    }
+    conf["resources"]["volumes"] = {
+        "count": args.volumes,
+        "size": 1,
+    }
+    conf["resources"]["networks"] = {
+        "count": args.networks,
+        "external_gateway": "public",
+    }
+    return conf
+
+
+def _get_session(conf):
+    auth = v3.Password(
+        auth_url=conf["auth_url"],
+        username=conf["username"],
+        password=conf["password"],
+        project_name=conf["project_name"],
+        user_domain_id=conf["user_domain_id"],
+        project_domain_id=conf["project_domain_id"],
+    )
+
+    return session_.Session(auth=auth, verify=True)
+
+
+def session_caching_wrapper():
+    cached = {}
+
+    def get_session():
+        if not cached.get("session"):
+            conf = get_conf()
+            cached["session"] = _get_session(conf["auth"])
+        return cached["session"]
+    return get_session
+
+
+get_session = session_caching_wrapper()
```

Comment:
I'm not sure what specifically you are querying here. `session_caching_wrapper()` returns a closure, this allows `get_session` to be called whenever a session is required and so saves having to pass a session object around.